Upgrade to use Node.js 20 and bump npm dependencies

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

Closes #2002

This PR:
- Bump npm dependencies and upgrade to use Node.js v20.
- Fix the build config of Fast Refresh mode.
- Address Sass deprecation warnings.
   - Ref: https://sass-lang.com/documentation/breaking-changes/mixed-decls/
- Turn off a few eslint rules:
   - `react/react-in-jsx-scope` rule
   - `jsdoc/check-line-alignment` rule
   - `import/no-named-as-default` rule
- Resolve eslint errors: (💡 It would be easier to review the changes by individual commits)
- Automatically fixed errors by running `npm run lint:js -- --fix`: [c41ef64](https://github.com/woocommerce/google-listings-and-ads/pull/2517/commits/c41ef64f1b2a3e3cef675729185d8797fded520b)
- `@wordpress/i18n-text-domain` rule: [cdd9aab](https://github.com/woocommerce/google-listings-and-ads/pull/2517/commits/cdd9aabfd374206cea3b525875f4733ccf95c8b7)
- `jsdoc/no-undefined-types` rule: [67ef0e2](https://github.com/woocommerce/google-listings-and-ads/pull/2517/commits/67ef0e2225dd452423708c6c62a23946d8bb0280)
- `import/no-unresolved` rule: [c4db2a3](https://github.com/woocommerce/google-listings-and-ads/pull/2517/commits/c4db2a3f4dd55075bc06d70e0b98beb5eec2a1b7)
- `no-redeclare` rule: [6997b6f](https://github.com/woocommerce/google-listings-and-ads/pull/2517/commits/6997b6fa2730e0be5c507e2f556710c4d3c8c066)
- `jsdoc/require-yields-check` rule: [47d1f9f](https://github.com/woocommerce/google-listings-and-ads/pull/2517/commits/47d1f9f44fe466171456a131d0ec74290d1c7d6f)
- `@typescript-eslint/no-use-before-define` rule: [9886344](https://github.com/woocommerce/google-listings-and-ads/pull/2517/commits/9886344a7fea97645b420481271d0e5a8952a128)
- `jsdoc/tag-lines` rule: [fccc0cd](https://github.com/woocommerce/google-listings-and-ads/pull/2517/commits/fccc0cde1a9e21f1ddd9de9c5805895b7b080eae)

❗ Since the jest testing requires a lot of adjustments, it will be handled by #2544. Two of the eslint rule errors are highly related to jest and are therefore handled by #2544 as well.

### Detailed test instructions:

1. Set Node.js to v20
2. Run `npm install` to see if it can finish without errors
3. Check if the following scripts work well
   - `npm run build`
   - `npm run dev`
   - `npm run start`
   - `npm run start:hot`
   - `npm run lint:css`
   - `npm run lint:js`
      - There are two pre-existing warnings that still can not be fixed after this upgrade
      - Errors from the `testing-library/no-unnecessary-act` and `testing-library/no-unnecessary-act` rules are handled separately in #2544
4. Run `node ./node_modules/.bin/bundlewatch` to check if the check works
5. Do some smoke testing

### Changelog entry

> Dev - Upgrade to use Node.js 20 and bump npm dependencies.

eason9487 committed:
Fix the build config of Fast Refresh mode.

eason9487 committed:
Turn off the eslint rule `react/react-in-jsx-scope`.

eason9487 committed:
Resolve eslint errors for the `@wordpress/i18n-text-domain` rule.

eason9487 committed:
Address Sass deprecation warnings.

Ref: https://sass-lang.com/documentation/breaking-changes/mixed-decls/

eason9487 committed:
Turn off the eslint rule `jsdoc/check-line-alignment`.

eason9487 committed:
Resolve eslint errors that can be automatically fixed.

Fixed by running `npm run lint:js -- --fix`.

eason9487 committed:
Resolve eslint errors for the `jsdoc/no-undefined-types` rule.

eason9487 committed:
Resolve eslint errors for the `import/no-unresolved` rule.

eason9487 committed:
Resolve eslint errors for the `no-redeclare` rule.

eason9487 committed:
Resolve eslint errors for the `jsdoc/require-yields-check` rule.

eason9487 committed:
Resolve eslint errors for the `@typescript-eslint/no-use-before-define` rule.

eason9487 committed:
Turn off the eslint rule `import/no-named-as-default`.

eason9487 committed:
Bump npm dependencies and upgrade to use Node.js v20.

eason9487 committed:
Resolve eslint errors for the `jsdoc/tag-lines` rule.

puntope commented:
> Address Sass deprecation warnings.

I still see the deprecations in both this PR and in https://github.com/woocommerce/google-listings-and-ads/pull/2544

<img width="1541" alt="Screenshot 2024-08-21 at 14 04 13" src="https://github.com/user-attachments/assets/b2e037ba-fe15-41e8-93f3-d97a121cca59">

eason9487 commented:
Hi @puntope

> I still see the deprecations in both this PR and in #2544

The SCSS styles that caused deprecations came from `@wordpress/base-styles` and could not be resolved from this repo.

puntope commented on js/src/components/contact-information/contact-information-preview-card.scss:
```diff
@@ -13,9 +13,9 @@
 	&__placeholder {
-		@include placeholder;
-
 		display: inline-block;
 		width: 18em;
+
+		@include placeholder;
```
Is this change necessary?  I wonder if placeholder includes  `display` or `width` props. Then it will be overrided

eason9487 replied:
> Is this change necessary?

The change fixes a Sass deprecation warning. The explanation can be found in https://sass-lang.com/documentation/breaking-changes/mixed-decls/

![image](https://github.com/user-attachments/assets/49861298-ba96-4344-be87-0007085d1a03)

> I wonder if placeholder includes `display` or `width` props. Then it will be overrided

They don't have. Please refer to:

https://github.com/woocommerce/google-listings-and-ads/blob/fccc0cde1a9e21f1ddd9de9c5805895b7b080eae/js/src/css/abstracts/_mixins.scss#L9-L21

puntope commented on js/src/components/free-listings/setup-free-listings/index.js:
```diff
@@ -2,6 +2,7 @@
  * External dependencies
  */
 import { useRef } from '@wordpress/element';
+import { Form } from '@woocommerce/components';
```
Wondering if it's worth it to import the component just for the docs.

eason9487 replied:
The main difference is the code editor's reference hints will be better, apart from that there shouldn't be any difference. So it should be worth it.

**Without import**

![image](https://github.com/user-attachments/assets/0827ae3c-8504-44ff-9b32-b6f3af56ff43)

**With import**
![image](https://github.com/user-attachments/assets/a0fca62c-39ac-4b96-949c-405861174034)

puntope commented on package.json:
```diff
@@ -171,7 +170,7 @@
 		}
 	},
 	"engines": {
-		"node": "^16 || ^18",
-		"npm": "^8 || ^9"
```
So I assume dependabot now supports

"node": "^20", and "npm": "^10"

?

eason9487 replied:
Thanks for catching this!

I thought dependabot would use paired node and npm versions. Surprisingly, it's using unpaired Node.js 20 and npm 9. Even the npm version claimed by its error message is incorrect, bringing more confusion.

```
Dependabot uses Node.js v20.17.0 and NPM 10.8.2. Due to the engine-strict setting, the update will not succeed.
```

Fixed in https://github.com/woocommerce/google-listings-and-ads/pull/2517/commits/b2a3d130e3a54cdb3fed71832962ca2ea6e10468

puntope commented on .eslintrc.js:
```diff
@@ -39,19 +40,78 @@ module.exports = {
 		],
+		// Turn it off temporarily because it involves a lot of re-alignment. We can revisit it later.
+		'jsdoc/check-line-alignment': 'off',
+		// The JS package `tracking-jsdoc` changes the definition of the `@fires` tag.
```
Can you elaborate what do you mean by this? 

 The JS package `tracking-jsdoc` changes the definition of the `@fires` tag.

eason9487 replied:
Adjusted the code comment in https://github.com/woocommerce/google-listings-and-ads/pull/2517/commits/c76514babdc1a4f51eab7b390b75aba61083d586. Let me know if it's still not clear enough.

puntope replied:
I will just write "Listing our event names to avoid eslint false alarms"

But not a blocker tho.

puntope reviewed (commented):
Howdy @eason9487 I tested all the scripts, code and did several smoke testing:

- Onboarding
- Edit campaigns
- Create rules in attribute mapping
- Campaign assets

Everything is working fine without errors so far.

I just left some comments regarding one CSS and just to be sure dependabot is compatible with the new node version.

eason9487 committed:
Allow npm `^9` because dependabot is using an npm version that doesn't pair with Node.js 20.

Address: https://github.com/woocommerce/google-listings-and-ads/pull/2517#discussion_r1726695456

eason9487 committed:
Add a more detailed code comment to the `jsdoc/no-undefined-types` eslint rule.

Ref: https://github.com/woocommerce/google-listings-and-ads/pull/2517#discussion_r1726709039

eason9487 commented:
Hi @puntope, thanks for the review! This PR is ready for a new round of code reviews.

puntope reviewed (approved):
✅  LGTM 

Thanks for the changes and all your work here
